clesperanto/core/execute.py

```diff
@@ -1,125 +1,108 @@
 
+    :param anchor: Enter __file__ when calling this method and the corresponding open.cl
+                   file lies in the same folder as the python file calling it.
+    :param kernel_name: kernel method inside the open.cl file to be called
+    :param global_size: global_size according to OpenCL definition (usually size of the
+                        destination image.)
+    :param parameters: dictionary containing parameters. Take care: They must be of the
+                       right type
+    :return:
+    """
 
-    if (len(global_size) == 2):
-        global_size = (global_size[1], global_size[0])
-    else:
-        global_size = (global_size[2], global_size[1], global_size[0])
+    defines = [
+        "#define GET_IMAGE_WIDTH(image_key) IMAGE_SIZE_ ## image_key ## _WIDTH",
+        "#define GET_IMAGE_HEIGHT(image_key) IMAGE_SIZE_ ## image_key ## _HEIGHT",
+        "#define GET_IMAGE_DEPTH(image_key) IMAGE_SIZE_ ## image_key ## _DEPTH",
+    ]
 
-    #print("global_size")
-    #print(global_size)
+    arguments = []
 
-    prog = OCLProgram(src_str=ocl_code)
-    prog.run_kernel(kernel_name, global_size, None, *arguments) # Todo: the order of the arguments matters; fix that
+    for key, value in parameters.items():
 
+        if isinstance(value, cl.array.Array):
+            arguments.append(value.data)
 
+            if value.dtype != np.dtype("float32"):
+                raise TypeError(
+                    "Only float32 is currently supported for images/buffers/arrays"
+                )
+
+            # image type handling
+            typeId = "f"
+            ndim = value.ndim
+            depth_height_width = [1, 1, 1]
+            depth_height_width[-len(value.shape) :] = value.shape
+            depth, height, width = depth_height_width
+            pos_type = "int2" if ndim < 3 else "int4"
+            pos = ["(pos0, 0)", "(pos0, pos1)", "(pos0, pos1, pos2, 0)"][ndim - 1]
+            img_dims = 2 if ndim < 3 else 2
+
+            defines.extend(
+                [
+                    f"#define CONVERT_{key}_PIXEL_TYPE clij_convert_float_sat",
+                    f"#define IMAGE_{key}_TYPE __global float*",
+                    f"#define IMAGE_{key}_PIXEL_TYPE float",
+                    f"#define IMAGE_SIZE_{key}_WIDTH {width}",
+                    f"#define IMAGE_SIZE_{key}_HEIGHT {height}",
+                    f"#define IMAGE_SIZE_{key}_DEPTH {depth}",
+                    f"#define POS_{key}_TYPE {pos_type}",
+                    f"#define POS_{key}_INSTANCE(pos0,pos1,pos2,pos3) ({pos_type}){pos}",
+                    f"#define READ_{key}_IMAGE(a,b,c) read_buffer{img_dims}d{typeId}"  # !
+                    "(GET_IMAGE_WIDTH(a),GET_IMAGE_HEIGHT(a),GET_IMAGE_DEPTH(a),a,b,c)",
+                    f"#define WRITE_{key}_IMAGE(a,b,c) write_buffer{img_dims}d{typeId}"  # !
+                    "(GET_IMAGE_WIDTH(a),GET_IMAGE_HEIGHT(a),GET_IMAGE_DEPTH(a),a,b,c)",
```

Comment:
I feel like this should be a string, e.g.

```python
IMAGE_HEADER = """
#define CONVERT_{key}_PIXEL_TYPE clij_convert_float_sat
#define IMAGE_{key}_TYPE __global float*
...
"""
```

Then this would be 

```python
defines.extend(
    IMAGE_HEADER.format(**locals()).split('\n')
)

Reply:
done